Fer va estimation

negarhdr opened the pull request:
This PR adds image-based facial expression recognition and valence-arousal estimation. 
It includes learner, unit-tests, demo, document, and ROS node. 
This is a replacement for the previous PR which had conflicts with other tools.

passalis commented on src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/dependencies.ini:
```diff
@@ -0,0 +1,14 @@
+       imutils
+       dlib
+
+#todo: update dependencies
```
Please remove this line, unless we need to update something.

passalis replied:
Please remove if not necessary.

passalis commented on projects/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
```diff
@@ -0,0 +1,96 @@
+# Skeleton-based Human Action Recognition Demo
```
Is the title and description correct?

passalis replied:
Please check the title again. Also, should this called image_based_facial_expression_recognition or ensemble_based_facial_expression_recognition? Both are used, which can be a bit confusing

passalis commented on src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py:
```diff
@@ -0,0 +1,719 @@
+            overall_dimension_preds += out_va_eval_cpu
+        ensemble_dimension_results = overall_dimension_preds / self.ensemble_size
+
+        return ensemble_emotion_results, ensemble_dimension_results
```
Please check existing data types (e.g., [Category](https://github.com/opendr-eu/opendr/blob/master/src/opendr/engine/target.py#L105) ) to check if they can be re-used in order to be consistent with the way the rest of the tools work.

negarhdr replied:
Infer returns two outputs. One is ensemble_emotion_results which is categorical emotion predictions and it is of data type Category. The other is ensemble_dimension_results which is valence-arousal continuous values and it is returned just as real numbers.

passalis reviewed (dismissed):
Thank you! I have added a few comments directly on files.

Furthermore, there are few additional points:

- Please make sure that you use the conventions for ROS argument passing (check also the names of the topics and arguments in https://github.com/opendr-eu/opendr/pull/281)
- Should this file be empty src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md ?
- The demo is nice. However, for most other tools for inference demo, we provide small examples of how each tool can be used. This demo incorporates too much functionality (which is good, but it can be also hard for new users to follow). So I would propose to also include a small inference demo that runs on a single image and clearly imports and uses the learner (similar to other tools, e.g., https://github.com/opendr-eu/opendr/blob/master/projects/perception/lightweight_open_pose/demos/inference_demo.py). In this way it would be clear for new users how they should use the tool.

ad-daniel commented:
The reason this PR keeps failing is because the algorithm puts a strict constraint on the dependency version of torch and torchvision [here](https://github.com/opendr-eu/opendr/blob/fer_va_estimation/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/dependencies.ini). 
We target `torch==1.9.0` and `torchvision==0.10.0`. This causes conflicts with other tools. Have you tried if it works anyway with less restrictive requirements? Or isn't there perhaps a more recent version of the algorithm?

Other dependencies might also be problematic, I've just checked those two

tsampazk commented:
By taking a quick look at the contents of this PR, i noticed that it uses the old directory structure in the projects folder. In essence, the `facial_expression_recognition` folder is now in `projects/*python*/perception/` instead of `projects/perception/`, so it would be helpful to move the `image_based_facial_emotion_estimation` folder that is contained in the new location.

Moreover, when trying to run the demo i noticed that the download method of the learner does not provide a way to download a pretrained model, but only test data (that i could not locate on the FTP, should we add it?). Will we provide a pretrained model for this tool? It would be helpful to have a pretrained model even internally to test the functionality of the tool.

negarhdr commented:
> By taking a quick look at the contents of this PR, i noticed that it uses the old directory structure in the projects folder. In essence, the `facial_expression_recognition` folder is now in `projects/*python*/perception/` instead of `projects/perception/`, so it would be helpful to move the `image_based_facial_emotion_estimation` folder that is contained in the new location.
> 
> Moreover, when trying to run the demo i noticed that the download method of the learner does not provide a way to download a pretrained model, but only test data (that i could not locate on the FTP, should we add it?). Will we provide a pretrained model for this tool? It would be helpful to have a pretrained model even internally to test the functionality of the tool.

Thanks for your comment and review. 
I updated the directory and also removed all the restrictions on the dependencies versions. 
The download method is also updated and it can download both pretrained model and data. However, it is now on Owncloud and needs to be moved to FTP server (I asked Pavlos to do so). However, it is still keep failing the tests. Are there any conflicts with other tools regarding a specific dependency?

ad-daniel commented:
Hi @negarhdr, I've fixed the problem with the dependency. Now the other tools are working but the `facial_expression_recognition` test appears to be failing (which was changed in this PR). Since the dependencies used by the algorithm were very old, you might need to do additional changes and enforce specific versions for the dependencies. If required to, try to stay as close as possible to the versions used by other tools to avoid conflicts. Hopefully you won't need to.
Could you fix the remaining problems and put the PR back to ready for review once done?

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  List of GPU indices to be used if the device is 'cuda'.
+- **validation_interval**: *int, default=1*\
+  Specifies the validation interval.
+- **max_training_epoch**: *int, default=100*\
```
```suggestion
- **max_training_epoch**: *int, default=2*\
```

ad-daniel replied:
according to learner

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  String name for saving checkpoints.
+- **dimensional_finetune**: *bool, default=True*\
+  Specifies if the model should be fintuned on dimensional data or not.
+- **categorical_train**: *bool, default=True*\
```
```suggestion
- **categorical_train**: *bool, default=False*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  Specifies if the model should be trained on categorical data or not.
+- **base_path_to_dataset**: *str, default=''./data/AffectNet''*\
+  Specifies the dataset path.
+- **max_tuning_epoch**: *int, default=20*\
```
```suggestion
- **max_tuning_epoch**: *int, default=1*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  Specifies the dataset path.
+- **max_tuning_epoch**: *int, default=20*\
+  Specifies the maximum number of epochs the model should be finetuned on dimensional data.
+- **max_tuning_epoch**: *bool, default=False*\
```
```suggestion
- **diversity**: *bool, default=False*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+
+Parameters:
+
+- **num_branches**: *int, default=9*\
```
```suggestion
- **num_branches**: *int*\
```

ad-daniel replied:
no default value on the learner side

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
````diff
@@ -0,0 +1,288 @@
+
+#### `FacialEmotionLearner.save`
+```python
+FacialEmotionLearner.save(self, state_dicts, base_path_to_save_model, current_branch_save)
````
```suggestion
FacialEmotionLearner.save(self, state_dicts, base_path_to_save_model)
```

ad-daniel replied:
no such parameter

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
````diff
@@ -0,0 +1,288 @@
+FacialEmotionLearner.save(self, state_dicts, base_path_to_save_model, current_branch_save)
+```
+This method is used to save a trained model.
+Provided with the path (absolute or relative), it creates the "path" directory, if it does not already
+exist.
````
```suggestion
Provided with the path (absolute or relative), it creates the "path" directory, if it does not already exist.
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  Object of type Python dictionary containing the trained model weights.
+- **base_path_to_save_model**: *str*\
+  Specifies the path in which the model will be saved.
+- **current_branch_save**: *int*\
+  Specifies the index of the model branch which should be saved.
```
```suggestion
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+
+- **ensemble_size**: *int, default=9*\
+  Specifies the number of ensemble branches in the model for which the pretrained weights should be loaded.
+- **path_to_saved_network**: *str*\
```
```suggestion
- **path_to_saved_network**: *str, default="./trained_models/esr_9"*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  Specifies the number of ensemble branches in the model for which the pretrained weights should be loaded.
+- **path_to_saved_network**: *str*\
+  Path of the model to be loaded.
+- **file_name_base_network**: *str*\
```
```suggestion
- **file_name_base_network**: *str, default="Net-Base-Shared_Representations.pt"*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  Path of the model to be loaded.
+- **file_name_base_network**: *str*\
+  The file name of the base network to be loaded.
+- **file_name_conv_branch**: *str*\
```
```suggestion
- **file_name_conv_branch**: *str, default="Net-Branch_{}.pt"*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+  The file name of the base network to be loaded.
+- **file_name_conv_branch**: *str*\
+  The file name of the ensemble branch network to be loaded.
+- **fix_backbone**: *bool, default=True*\
```
```suggestion
- **fix_backbone**: *bool*\
```

ad-daniel commented on docs/reference/image_based_facial_emotion_estimation.md:
```diff
@@ -0,0 +1,288 @@
+- **name_experiment**: *str, default='esr_9'*\
+  String name for saving checkpoints.
+- **dimensional_finetune**: *bool, default=True*\
+  Specifies if the model should be fintuned on dimensional data or not.
```
```suggestion
  Specifies if the model should be fine-tuned on dimensional data or not.
```

ad-daniel commented on projects/opendr_ws/src/perception/scripts/facial_emotion_estimation.py:
```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python
```
The location for the ROS scripts has been changed, this file should be under: opendr/projects/opendr_ws/opendr_perception

tsampazk replied:
I moved it and also applied several fixes on the node, so i mark this as resolved @negarhdr

ad-daniel commented on projects/opendr_ws/src/perception/scripts/facial_emotion_estimation.py:
```diff
@@ -0,0 +1,203 @@
+        Start the node and begin processing input data
+        """
+        rospy.init_node('opendr_facial_emotion_estimation', anonymous=True)
+        rospy.Subscriber(self.input_image_topic, ROS_Image, self.callback)
```
```suggestion
        rospy.Subscriber(self.input_rgb_image_topic, ROS_Image, self.callback)
```

ad-daniel replied:
also, do we need to set queue and buffer size to avoid delays?

tsampazk replied:
Also fixed this @negarhdr .

ad-daniel commented on projects/opendr_ws/src/perception/scripts/facial_emotion_estimation.py:
```diff
@@ -0,0 +1,203 @@
+    # Select the device for running the
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-i', '--input_image_topic', type=str, help='Topic name for input rgb image',
```
```suggestion
    parser.add_argument('-i', '--input_rgb_image_topic', type=str, help='Topic name for input rgb image',
```

ad-daniel commented on projects/opendr_ws/src/perception/scripts/facial_emotion_estimation.py:
```diff
@@ -0,0 +1,203 @@
+        device = 'cpu'
+
+    facial_emotion_estimation_node = FacialEmotionEstimationNode(
+        input_image_topic=args.input_image_topic,
```
```suggestion
        input_rgb_image_topic=args.input_rgb_image_topic,
```

tsampazk replied:
Fixed @negarhdr

ad-daniel commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/benchmark_esr.py:
```diff
@@ -0,0 +1,94 @@
+# Copyright 2020-2022 OpenDR European Project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
```
A benchmark script is provided but no information seems to be present in the documentation, are there any results available yet?

ad-daniel commented on src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py:
```diff
@@ -0,0 +1,764 @@
+                 max_tuning_epoch=1, diversify=False
+                 ):
+        super(FacialEmotionLearner, self).__init__(lr=lr, batch_size=batch_size, temp_path=temp_path, device=device)
+        # dataset_name = ''
+        # device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
```suggestion
```

ad-daniel commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+Please note that the pretrained weights cannot be used for commercial purposes!
+To recognize a facial expression in images, run the following command:
+```python
+python inference_demo.py image -i ./media/jackie.jpg -pre ./pretrained -d -s 2
````
since these are the instructions on how to run the demo, I think we should include the image in the repo

tsampazk replied:
Preferably the image should be uploaded on the FTP and downloaded via the download method of the learner as "test_data".

negarhdr replied:
I uploaded a sample image and a sample video to the Owncloud which needs to be moved to the FTP server. In case the user does not provide an image/video path, it automatically downloads the sample image/video from the FTP server. 

The folder "demo" on Owncloud: 
`OpenDR/FTP Server Material/perception/facial_emotion_estimation/demo`

should move to FTP server: 
`perception/facial_emotion_estimation/`

ad-daniel commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+
+To recognize a facial expression in images captured from a webcam, run the following command:
+```python
+python inference_demo.py webcam -pre 'pretrained_path' -d -s 2 
````
When I try the webcam demo, I get this:

```
Traceback (most recent call last):
  File "inference_demo.py", line 394, in <module>
    main()
  File "inference_demo.py", line 386, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 210, in webcam
    fer_display = FERDisplay(screen_size=screen_size, display_graph_ensemble=(not no_plot))
  File "/home/daniel/opendr/projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/gui/fer_display.py", line 133, in __init__
    dtype=np.int)
  File "/home/daniel/opendr/venv/lib/python3.8/site-packages/numpy/__init__.py", line 284, in __getattr__
    raise AttributeError("module {!r} has no attribute "
AttributeError: module 'numpy' has no attribute 'int'
```
after installing the toolkit from this branch, just me?

ad-daniel replied:
This is likely unrelated to this PR, seeing the issue elsewhere. Might not be a bug.

tsampazk replied:
When i ran `python3 inference_demo.py webcam -d` i got the following error:

```
Error raised during video mode.
Traceback (most recent call last):
  File "inference_demo.py", line 394, in <module>
    main()
  File "inference_demo.py", line 386, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 236, in webcam
    raise e
  File "inference_demo.py", line 222, in webcam
    fer = None if (img is None) else recognize_facial_expression(img, device, gradcam,
  File "inference_demo.py", line 161, in recognize_facial_expression
    face_coordinates = detect_face(image)
  File "inference_demo.py", line 69, in detect_face
    faces = _FACE_DETECTOR_HAAR_CASCADE.detectMultiScale(greyscale_image, scaleFactor=1.2, minNeighbors=9,
cv2.error: OpenCV(4.5.1) /tmp/pip-req-build-ms668fyv/opencv/modules/objdetect/src/cascadedetect.cpp:1689: error: (-215:Assertion failed) !empty() in function 'detectMultiScale'
```

@negarhdr is it something wrong on my end, got any ideas? :)

negarhdr replied:
It should now download the pretrained weights automatically from the FTP server.

negarhdr replied:
> When i ran `python3 inference_demo.py webcam -d` i got the following error:
> 
> ```
> Error raised during video mode.
> Traceback (most recent call last):
>   File "inference_demo.py", line 394, in <module>
>     main()
>   File "inference_demo.py", line 386, in main
>     webcam(args.webcam_id, args.display, args.gradcam, args.output,
>   File "inference_demo.py", line 236, in webcam
>     raise e
>   File "inference_demo.py", line 222, in webcam
>     fer = None if (img is None) else recognize_facial_expression(img, device, gradcam,
>   File "inference_demo.py", line 161, in recognize_facial_expression
>     face_coordinates = detect_face(image)
>   File "inference_demo.py", line 69, in detect_face
>     faces = _FACE_DETECTOR_HAAR_CASCADE.detectMultiScale(greyscale_image, scaleFactor=1.2, minNeighbors=9,
> cv2.error: OpenCV(4.5.1) /tmp/pip-req-build-ms668fyv/opencv/modules/objdetect/src/cascadedetect.cpp:1689: error: (-215:Assertion failed) !empty() in function 'detectMultiScale'
> ```
> 
> @negarhdr is it something wrong on my end, got any ideas? :)

Hi, 
apparently it couldn't read the face_detector 'xml' file. I updated the path, and it should work now. Please let me know if there is the same problem in the ROS node as well.

ad-daniel reviewed (changes requested):
Thank you, only a few details

- Add changelog entry
- Link documentation page in index.md
- Link demo in index.md (towards the end of the page)

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
```diff
@@ -0,0 +1,96 @@
+
+#### Running demo
+The pretrained models on AffectNet Categorical dataset are provided by [[1]](#1) which can be found [here](https://github.com/siqueira-hc/Efficient-Facial-Feature-Learning-with-Wide-Ensemble-based-Convolutional-Neural-Networks/tree/master/model/ml/trained_models/esr_9).
+Please note that the pretrained weights cannot be used for commercial purposes!
```
Might be more eye-catching to make it bold.

```suggestion
**Please note that the pretrained weights cannot be used for commercial purposes**
```

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+python inference_demo.py image -i ./media/jackie.jpg -pre ./pretrained -d -s 2
+```  
+
+The argument "image" indicates that the input is an image. The location of the image is specified after -i while -pre indicates the location of pretrained model weights. -d sets the display mode to true and -s 2 sets the window size to 1440 x 900.
+You can also visualize regions in the image relevant for the classification of facial expression by adding -g as arguments:
````
Added some formatting on the arguments etc to better distinguish them.
```suggestion
The argument `image` indicates that the input is an image. The location of the image is specified after `-i` while `-pre` indicates the location of pretrained model weights. `-d` sets the display mode to true and `-s 2` sets the window size to 1440 x 900.
You can also visualize regions in the image relevant for the classification of facial expression by adding `-g` as arguments:
```

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+```python
+python inference_demo.py image -i 'image_path' -pre 'pretrained_path' -d -s 2 -g
+```  
+The argument -g generates saliency maps with the Grad-CAM algorithm.
````
```suggestion
The argument `-g` generates saliency maps with the Grad-CAM algorithm.
```

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+```python
+python inference_demo.py video -i 'video_path' -pre 'pretrained_path' -d -f 5 -s 2
+```
+The argument "video" indicates that the input is a video. The location of the video is specified after -i. -d sets the display mode to true, -f defines the number of frames to be processed, and -s 2 sets the window size to 1440 x 900.
````
```suggestion
The argument `video` indicates that the input is a video. The location of the video is specified after `-i`. `-d` sets the display mode to true, `-f` defines the number of frames to be processed, and `-s 2` sets the window size to 1440 x 900.
```

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/README.md:
````diff
@@ -0,0 +1,96 @@
+```python
+python inference_demo.py webcam -pre 'pretrained_path' -d -s 2 
+```
+The argument "webcam" indicates the framework to capture images from a webcam. -d sets the display mode to true, -s 2 sets the window size to 1440 x 900.
````
```suggestion
The argument `webcam` indicates the framework to capture images from a webcam. `-d` sets the display mode to true, `-s 2` sets the window size to 1440 x 900.
```

tsampazk commented on projects/python/perception/facial_expression_recognition/image_based_facial_emotion_estimation/gui/grad_cam.py:
```diff
@@ -0,0 +1,100 @@
+        (AAAI-20), pages 1–1, New York, USA.
+    """
+
+    # def __init__(self, esr_base, esr_branch_to_last_conv_layer, esr_branch_from_last_conv_layer_to_emotion_output):
```
This could probably be removed.

```suggestion
```

tsampazk reviewed (changes requested):
Thanks @negarhdr! I resolved the issues with the ROS node and ran it but i got the following error:
```
Traceback (most recent call last):
  File "/home/kostas/dev/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 192, in <module>
    facial_emotion_estimation_node = FacialEmotionEstimationNode(
  File "/home/kostas/dev/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 76, in __init__
    self.facial_emotion_estimator.load(ensemble_size=9, path_to_saved_network=model_saved_path)
  File "/home/kostas/dev/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py", line 149, in load
    self.model.base.load_state_dict(torch.load(
AttributeError: 'NoneType' object has no attribute 'base'
```

tsampazk commented on projects/opendr_ws/src/opendr_perception/scripts/facial_emotion_estimation_node.py:
```diff
@@ -0,0 +1,198 @@
+from sensor_msgs.msg import Image as ROS_Image
+from opendr_bridge import ROSBridge
+
+from opendr.engine.constants import OPENDR_SERVER_URL
```
Sorry Negar i missed this when i updated the node. This import is unused, could you remove it when you get the chance? I am not doing it myself to not mess up your branch.

```suggestion
```

negarhdr commented:
Are the failing tests related to my PR? As I see, it's object_detection and human_model_generation.

tsampazk commented:
> Are the failing tests related to my PR? As I see, it's object_detection and human_model_generation.

No these rate limit errors are unrelated, there's a fix coming as soon as #370 is merged.

tsampazk commented on src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py:
```diff
@@ -0,0 +1,804 @@
+            if not os.path.exists(model_path):
+                os.makedirs(model_path)
+
+            zip_path = os.path.join(path, 'pretrained/esr_9.zip')
+            if not os.path.exists(zip_path):
+                # Download data
+                file_url = os.path.join(url, 'pretrained/esr_9.zip')
+                urlretrieve(file_url, zip_path)
+                with zipfile.ZipFile(zip_path, 'r') as zip_ref:
+                    zip_ref.extractall(model_path)
+                os.remove(zip_path)
+            else:
+                print("Pretrained files already exist.")
```
I noticed that this code ends up re-downloading every time. The exists check is made for the zip file which is removed on line 766. It should probably check for the extracted directory or maybe it shouldn't remove the zip file, whichever you prefer.

Same applies for data files on line 741,etc. above.

negarhdr replied:
Thank you for the comment. I removed the zip remover.

tsampazk reviewed (changes requested):
Thank you for the fixes so far @negarhdr! 

Unfortunately, i still get an error from the ROS node:

```
Error on trying to read the following file as an image: [[[115 118 122 ...  61  64  72]
  [116 120 124 ...  62  63  68]
  [118 123 128 ...  64  62  63]
  ...
  [ 30  28  27 ...  50  50  49]
  [ 29  28  28 ...  50  50  48]
  [ 31  30  28 ...  49  49  48]]

 [[117 120 118 ...  72  75  83]
  [118 122 121 ...  71  74  79]
  [119 124 127 ...  73  73  74]
  ...
  [ 33  31  31 ...  57  56  55]
  [ 32  31  31 ...  57  57  55]
  [ 31  30  31 ...  56  56  55]]

 [[ 99 102 119 ...  68  71  79]
  [100 104 122 ...  68  70  75]
  [104 109 124 ...  70  69  70]
  ...
  [ 34  32  32 ...  51  54  53]
  [ 33  32  32 ...  51  51  49]
  [ 33  32  32 ...  50  50  49]]]
Please, check if the file exists, is an image and is not corrupted.
Supported file formats: JPEG (*.jpeg and *.jpg) and Portable Network Graphics (*.png).
[ERROR] [1671546034.769423]: bad callback: <bound method FacialEmotionEstimationNode.callback of <__main__.FacialEmotionEstimationNode object at 0x7f8d92e1f5b0>>
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "~/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 98, in callback
    image = image_processing.read(image)
  File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/algorithm/utils/image_processing.py", line 133, in read
    raise exception
  File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/algorithm/utils/image_processing.py", line 119, in read
    loaded_image = cv2.imread(path_to_image, cv2.IMREAD_COLOR)
TypeError: Can't convert object of type 'Image' to 'str' for 'filename'

```

I also got some issues with the inference demo, regarding the arguments of argparse that i went ahead and fixed, but ended up on this one:

```
Error raised during video mode.
Traceback (most recent call last):
  File "inference_demo.py", line 405, in <module>
    main()
  File "inference_demo.py", line 394, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 234, in webcam
    raise e
  File "inference_demo.py", line 221, in webcam
    fer = None if (img is None) else recognize_facial_expression(img, device, gradcam, ensemble_size)
  File "inference_demo.py", line 172, in recognize_facial_expression
    input_face = _pre_process_input_image(face)
  File "inference_demo.py", line 85, in _pre_process_input_image
    image = image_processing.resize(image, ESR.INPUT_IMAGE_SIZE)
AttributeError: type object 'ESR' has no attribute 'INPUT_IMAGE_SIZE'
```

by running `python3 inference_demo.py webcam -d -s 3`, as soon as a face was detected.

ad-daniel commented:
I can confirm the problems with the demo, even after Kostas' fixes I get the same error `AttributeError: type object 'ESR' has no attribute 'INPUT_IMAGE_SIZE'` when the face is detected (webcam mode)

negarhdr commented:
> Thank you for the fixes so far @negarhdr!
> 
> Unfortunately, i still get an error from the ROS node:
> 
> ```
> Error on trying to read the following file as an image: [[[115 118 122 ...  61  64  72]
>   [116 120 124 ...  62  63  68]
>   [118 123 128 ...  64  62  63]
>   ...
>   [ 30  28  27 ...  50  50  49]
>   [ 29  28  28 ...  50  50  48]
>   [ 31  30  28 ...  49  49  48]]
> 
>  [[117 120 118 ...  72  75  83]
>   [118 122 121 ...  71  74  79]
>   [119 124 127 ...  73  73  74]
>   ...
>   [ 33  31  31 ...  57  56  55]
>   [ 32  31  31 ...  57  57  55]
>   [ 31  30  31 ...  56  56  55]]
> 
>  [[ 99 102 119 ...  68  71  79]
>   [100 104 122 ...  68  70  75]
>   [104 109 124 ...  70  69  70]
>   ...
>   [ 34  32  32 ...  51  54  53]
>   [ 33  32  32 ...  51  51  49]
>   [ 33  32  32 ...  50  50  49]]]
> Please, check if the file exists, is an image and is not corrupted.
> Supported file formats: JPEG (*.jpeg and *.jpg) and Portable Network Graphics (*.png).
> [ERROR] [1671546034.769423]: bad callback: <bound method FacialEmotionEstimationNode.callback of <__main__.FacialEmotionEstimationNode object at 0x7f8d92e1f5b0>>
> Traceback (most recent call last):
>   File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
>     cb(msg)
>   File "~/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 98, in callback
>     image = image_processing.read(image)
>   File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/algorithm/utils/image_processing.py", line 133, in read
>     raise exception
>   File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/algorithm/utils/image_processing.py", line 119, in read
>     loaded_image = cv2.imread(path_to_image, cv2.IMREAD_COLOR)
> TypeError: Can't convert object of type 'Image' to 'str' for 'filename'
> ```
> 
> I also got some issues with the inference demo, regarding the arguments of argparse that i went ahead and fixed, but ended up on this one:
> 
> ```
> Error raised during video mode.
> Traceback (most recent call last):
>   File "inference_demo.py", line 405, in <module>
>     main()
>   File "inference_demo.py", line 394, in main
>     webcam(args.webcam_id, args.display, args.gradcam, args.output,
>   File "inference_demo.py", line 234, in webcam
>     raise e
>   File "inference_demo.py", line 221, in webcam
>     fer = None if (img is None) else recognize_facial_expression(img, device, gradcam, ensemble_size)
>   File "inference_demo.py", line 172, in recognize_facial_expression
>     input_face = _pre_process_input_image(face)
>   File "inference_demo.py", line 85, in _pre_process_input_image
>     image = image_processing.resize(image, ESR.INPUT_IMAGE_SIZE)
> AttributeError: type object 'ESR' has no attribute 'INPUT_IMAGE_SIZE'
> ```
> 
> by running `python3 inference_demo.py webcam -d -s 3`, as soon as a face was detected.

Thanks for the comment. I fixed the bugs. I actually cannot run the ROS node on our servers, and that's why I didn't notice the bugs.

tsampazk commented:
There was still a minor issue with how device argument was handled which i fixed. Still got errors on both the ROS node and the demo.

# For the ROS node:
```
[ERROR] [1671548909.918525]: bad callback: <bound method FacialEmotionEstimationNode.callback of <__main__.FacialEmotionEstimationNode object at 0x7f22ac6f45b0>>
Traceback (most recent call last):
  File "/opt/ros/noetic/lib/python3/dist-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "~/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 99, in callback
    face_coordinates = detect_face(image)
  File "~/opendr/projects/opendr_ws/install/lib/opendr_perception/facial_emotion_estimation_node.py", line 129, in detect_face
    greyscale_image = image_processing.convert_bgr_to_grey(image)
  File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/algorithm/utils/image_processing.py", line 156, in convert_bgr_to_grey
    return cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
TypeError: Expected Ptr<cv::UMat> for argument 'src'
```

# For the demo on --device cuda:
```
Error raised during video mode.
Traceback (most recent call last):
  File "~/opendr/src/opendr/engine/data.py", line 237, in __init__
    data = np.asarray(data)
  File "~/.local/lib/python3.8/site-packages/torch/_tensor.py", line 643, in __array__
    return self.numpy()
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "inference_demo.py", line 409, in <module>
    main()
  File "inference_demo.py", line 398, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 238, in webcam
    raise e
  File "inference_demo.py", line 225, in webcam
    fer = None if (img is None) else recognize_facial_expression(img, device, gradcam, ensemble_size)
  File "inference_demo.py", line 181, in recognize_facial_expression
    emotion, affect, emotion_idx, model = _predict(input_face, device, ensemble_size)
  File "inference_demo.py", line 115, in _predict
    emotion, affect = learner.infer(input_face)
  File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py", line 627, in infer
    input_batch = Image(input_batch)
  File "~/opendr/src/opendr/engine/data.py", line 239, in __init__
    raise ValueError("Image data not understood (cannot be cast to NumPy array).")
ValueError: Image data not understood (cannot be cast to NumPy array).
```

# For the demo on --device cpu:
```
Error raised during video mode.
Traceback (most recent call last):
  File "inference_demo.py", line 409, in <module>
    main()
  File "inference_demo.py", line 401, in main
    webcam(args.webcam_id, args.display, args.gradcam, args.output,
  File "inference_demo.py", line 238, in webcam
    raise e
  File "inference_demo.py", line 225, in webcam
    fer = None if (img is None) else recognize_facial_expression(img, device, gradcam, ensemble_size)
  File "inference_demo.py", line 181, in recognize_facial_expression
    emotion, affect, emotion_idx, model = _predict(input_face, device, ensemble_size)
  File "inference_demo.py", line 115, in _predict
    emotion, affect = learner.infer(input_face)
  File "~/opendr/src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py", line 627, in infer
    input_batch = Image(input_batch)
  File "~/opendr/src/opendr/engine/data.py", line 242, in __init__
    raise ValueError("3-dimensional images are expected")
ValueError: 3-dimensional images are expected
```

For both devices its the same line that produces the errors.

tsampazk commented:
> I actually cannot run the ROS node on our servers, and that's why I didn't notice the bugs.

It's ok i can test it the ROS node then. Is it possible to test the demo with a webcam and make sure it runs properly?

negarhdr commented:
> > I actually cannot run the ROS node on our servers, and that's why I didn't notice the bugs.
> 
> It's ok i can test the ROS node then. Is it possible to test the demo with a webcam and make sure it runs properly?

Yes, I test the demo and fix it.

tsampazk commented:
@negarhdr Thank you for the fixes! I fixed some minor stuff to make the demo work with a webcam, and it works ok now.

As for the ROS node, let me look into it for a bit and i'll get back to you.

tsampazk commented:
@ad-daniel @negarhdr The ROS1 node now uses retina face, as the cv face detector proved problematic with the conversions required for ROS. I also included the publishing of an annotated image which draws bounding boxes with all facial emotion info printed on the corner of the box. I tested it with multiple faces and it works well. It is similar to the face recognition pipeline in the corresponding node. Note that it is quite slow, but the detector is far superior in the performed qualitative testing.

ad-daniel reviewed (approved):
When running the ROS node I'm getting this:
```
rosrun opendr_perception facial_emotion_estimation_node.py
[rospack] Error: package 'opendr_perception' not found
find: ‘’: No such file or directory
Tokenization took: 0.01 ms
```
however it doesn't prevent it from working, so not a big deal.

As for the demo, I've tested the webcam and it works nicely. Image and video are failing, from the error it seems to be related to the fact it cannot find the resources in the FTP server. @Pavlos-Tosidis has it been moved from owncloud? I think we forgot to ping him

Other than that, it looks good to me, thank you both!

tsampazk commented:
> When running the ROS node I'm getting this:
> 
> ```
> rosrun opendr_perception facial_emotion_estimation_node.py
> [rospack] Error: package 'opendr_perception' not found
> find: ‘’: No such file or directory
> Tokenization took: 0.01 ms
> ```

Hmm that's weird, never gotten this error before. Maybe [this](https://stackoverflow.com/questions/56139647/how-to-fix-rospack-error-package-my-package-not-found) helps?

Thanks!

ad-daniel commented:
Nevermind my bad, I did a `source devel/local_setup.bash` rather than `source devel/setup.bash`.

tsampazk commented on src/opendr/perception/facial_expression_recognition/image_based_facial_emotion_estimation/facial_emotion_learner.py:
```diff
@@ -0,0 +1,799 @@
+            if not os.path.exists(demo_path):
+                os.makedirs(demo_path)
+
+            vid_path = os.path.join(demo_path, 'sheldon.mp4')
```
Typo in the filename, files should be up now.

```suggestion
            vid_path = os.path.join(demo_path, 'sheldon.jpg')
```

tsampazk replied:
sorry basically this is in the `demo_video` mode, it should probably set it to 'big_bang.mp4'

Pavlos-Tosidis commented:
> 

Truth is I never saw this comment for moving the demo on the FTP. Files are up now!

tsampazk commented:
Went ahead and resolved the issue with the path in download. Also fixed the demo for image and video modes, as the cv2 operations to show the images were used incorrectly. Should be good to go now!

ad-daniel reviewed (approved):
Working for me as well, thank you!

tsampazk commented:
Sorry Daniel just merged hr pose estimation and there was a conflict.
